File: assignment4/ClientUDP.java
import java.net.*;  // for DatagramSocket, DatagramPacket, and InetAddress
import java.io.*;   // for IOException
import java.util.Scanner;

public class ClientUDP {

   public static void main(String args[]) throws Exception {
   
      if (args.length != 2 && args.length != 3) // Test for correct # of args
         throw new IllegalArgumentException("Parameter(s): <Destination>" +
                                                   " <Port> [<encoding]");
   
   
      InetAddress destAddr = InetAddress.getByName(args[0]); // Destination address
      int destPort = Integer.parseInt(args[1]) + RequestBinConst.GROUP_NUMBER; // Destination port
      Byte ID = 1;
      Byte opCode;
      Short op1 = 0;
      Short op2 = 0;
   
      for(;;) {
         Byte TML = 6;
         Byte operands = 1;
      
      
         Scanner scan = new Scanner(System.in);
         System.out.println("Please Enter The Following Values:");
         System.out.print("\tOperand Type - \n\t\"0\" = +, \n\t\"1\" = -, \n\t\"2\" = *, \n\t\"3\" = /, \n\t\"4\" = >>, \n\t\"5\" = <<, \n\t\"6\" = ~, \n\t\"exit\" = Terminate \n\tEnter Selection: ");
      
         String opString = scan.nextLine();
      
         if (opString.equals("exit")) {
            break;
         }
         int opCodeInt = Integer.parseInt(opString);
         opCode = (byte) opCodeInt;
      
      
      
         System.out.print("\n\tOperand 1: ");
         int op1Int = Integer.parseInt(scan.nextLine());
         op1 = (short) op1Int;
      
      
         if (opCode <= 5) {
            System.out.print("\n\tOperand 2: ");
            int op2Int = Integer.parseInt(scan.nextLine());
            op2 = (short) op2Int;
            TML = (byte) 8;
            operands = (byte) 2;
         }
      
         Request Request = new Request(TML, ID, opCode, operands, op1, op2);
      
         DatagramSocket sock = new DatagramSocket(); // UDP socket for sending
      
      
                // Use the encoding scheme given on the command line (args[2])
         RequestEncoder encoder = (args.length == 3 ?
                                          new RequestEncoderBin(args[2]) :
                                          new RequestEncoderBin());
      
         byte[] codedRequest = encoder.encode(Request); // Encode Request
      
         DatagramPacket message = new DatagramPacket(codedRequest, codedRequest.length,
                                                            destAddr, destPort);
      
         System.out.println("\nMessage length: " + codedRequest.length);
         System.out.println("\nRequest Hex String:");
         byte[] buffer = message.getData();
         for (int i = 0; i < message.getLength(); i++) {
            System.out.format("\t0x%x\n", buffer[i]);
         }
      
      
         sock.send(message);
         long start = System.nanoTime();
      
         sock.receive(message);
         long finish = System.nanoTime();
         long elapsed = finish - start;
         RequestDecoder decoder = (args.length == 2 ? // Which encoding
                                          new RequestDecoderBin(args[1]) :
                                          new RequestDecoderBin() );
      
         Response Response = decoder.decodeResponse(message);
      
      
         System.out.print("Response Hex String: \n");
         byte[] byteBuffer = message.getData();
         for (int i = 0; i < Response.TML; i++) {
            System.out.format("\t0x%x\n", byteBuffer[i]);
         }
         System.out.println("\n" + Response);
         System.out.println("Time Elapsed: " + elapsed / 1000000.0 + "ms\n");
         ID++;
      }
   
   
        //sock.close();
   }
}

File: assignment4/ClientUDP.py
import socket
import sys
import time

GROUP_NUMBER = 24
DEFAULT_ENCODING = 'iso-8859-1'
MAX_WIRE_LENGTH  = 1024

ipAddress = sys.argv[1]
port = int(sys.argv[2])
gPort = port + GROUP_NUMBER

if len(sys.argv) != 2 and len(sys.argv) != 3:
    raise ValueError('Parameter(s): <Destination>  <Port> [<encoding]')

serverAddress = (ipAddress, gPort)

ID = 1
opCode = 1
op1 = 0
op2 = 0
exitStr = 'exit'

while (True):
    TML = 6
    operands = 1

    print("Please Enter The Following Values:\n")
    opString = input('\tOperand Type - \n\t\"0\" = + \n\t\"1\" = - \n\t\"2\" = * \n\t\"3\" = / \n\t\"4\" = >>  \n\t\"5\" = << \n\t\"6\" = ~ \n\t\"exit\" = Terminate \n\tEnter Selection: ')
    if str(opString) in "exit":
        sys.exit(0)

    opCode = int(opString)
    op1 = int(input("\n\tOperand 1: "))

    if opCode <= 5:
        op2 = int(input("\n\tOperand 2: "))
        TML = 8
        operands = 2

    requestStr = "TML: %d\nRequestID: %d\nOp Code: %d\nOperands: %d\nOperand 1: %d\nOperand 2: %d\n" % (
        TML, ID, opCode, operands, op1, op2)
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

    request = "%d%d%d%d%d%d" % (TML, ID, opCode, operands, op1, op2)
    codedRequest = bytearray()
    codedRequest.extend(request)



    print('\n\nMessage Length: %s' % TML)
    print('\n\nRequest Hex String: \n')
    # buffer = bytearray(codedRequest)
    for i in range(len(codedRequest)):
        print('\t0x%s\n' % int(codedRequest[i]))    #TODO LIST INDICES MUST BE INTEGERS, NOT STR
    send = sock.sendto(codedRequest, serverAddress)
    start = time.time()

    recv, server = sock.recvfrom(MAX_WIRE_LENGTH)
    finish = time.time()
    elapsed = start - finish

    response = [str(x).decode(DEFAULT_ENCODING) for x in recv]
    print('\n\nMessage Length: %d' % len(response))
    print('\n\nResponse Hex String: \n')
    byteBuffer = bytearray(recv)
    for i in recv:
        print('\t0x%s\n' % byteBuffer[i])
    print('\n%s' % response)
    final = elapsed / 1000000.0
    print('\n\nTime Elapsed: %dms\n' % final)
    ID = ID + 1
sock.close()
